### ISSUE TITLE
Jackson LockFreePool recycler pool causes memory issues

### ISSUE THREAD

**Issue by davseitsev** [trinodb/trino#21356]
After upgrade Trino from 409 to 443 we started to have memory issues. Some workers disappear from the cluster due to intensive major GC. In the logs we can see messages like:
```
INFO    Notification Thread     io.airlift.stats.JmxGcMonitor   Major GC: application 61ms, stopped 9353ms: 91.59GB -> 91.13GB
```
On the graphs we can see memory starvation:
![image](https://github.com/trinodb/trino/assets/1793410/ea3de493-3823-4dbd-b74c-ef6c90a9dd7e)

Heap dump analisys shows it comes from [JsonUtil#OBJECT_MAPPED_UNORDERED](https://github.com/trinodb/trino/blob/443/core/trino-main/src/main/java/io/trino/util/JsonUtil.java#L113). It creates `JsonFactory` with default implementation of `RecyclerPool` which is [LockFreePool](https://github.com/FasterXML/jackson-core/blob/jackson-core-2.17.0/src/main/java/com/fasterxml/jackson/core/util/JsonRecyclerPools.java#L29). This implementation is relatively young, has some concerns and is considered not to be used as default https://github.com/FasterXML/jackson-core/issues/1256

![image](https://github.com/trinodb/trino/assets/1793410/f1cc5e73-f58e-49c8-9bd3-0038359970d9)

We have change the factory below to use [JsonRecyclerPools.threadLocalPool()](https://github.com/FasterXML/jackson-core/blob/jackson-core-2.17.0/src/main/java/com/fasterxml/jackson/core/util/JsonRecyclerPools.java#L38)

https://github.com/trinodb/trino/blob/f0d5b0febd2c06fc79a63dec0c702a224e31c248/core/trino-main/src/main/java/io/trino/util/JsonUtil.java#L118-L121

I'm not sure it's the best implementation for Trino but now we can't reproduce the issue.

**Comment by davseitsev** [comment]
@hashhar I've applied your fix and testing it right now. I will update later.

**Comment by wendigo** [comment]
Closing then. Release should happen this week. Thanks!

**Comment by marvin-roesch** [comment]
This issue appears to still be affecting us on Trino 444. We are running into massive memory issues with this. In our case, heap dumps indicate that [`JsonTypeUtil.parseJson`](https://github.com/trinodb/trino/blob/master/lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/JsonTypeUtil.java#L46) is at fault.

Some quick debugging shows that the issue is due to the `SORTED_MAPPER.writeValue` call using the default `JsonGenerator` instance which in turn gets created by a default `JsonFactory` that uses the `LockFreePool` implementation. Until the upstream maybe changes back the default again, it might be needed to perform a more systematic survey of Jackson usages to make sure this doesn't negatively affect more places.

Ideally, all `ObjectMapper` instances would get created with the customized `JsonFactory` that contains the configuration.

Alternatively, maybe explicitly downgrading the Jackson libraries is an option?

**Comment by wendigo** [comment]
`SORTED_MAPPER` is using `ObjectMapperProvider` which configures the `JsonFactoryBuilder` to use `JsonRecyclerPools.threadLocalPool()`. if this is ineffective I guess we need to report Jackson issue.

**Comment by marvin-roesch** [comment]
That's great to hear, thanks! I wasn't checking other dependencies' upstream when investigating this. Will report back once 445 is out.

**Comment by mariofusco** [comment]
Note that we're discussing this issue also [here](https://github.com/FasterXML/jackson-core/issues/1260), so in case you have any suggestion on how to reproduce the problem it would be great to report it also there.

**Comment by wendigo** [comment]
@mariofusco have you seen how we are parsing JSON: https://github.com/trinodb/trino/blob/master/lib/trino-plugin-toolkit/src/main/java/io/trino/plugin/base/util/JsonTypeUtil.java#L46 ?

**Comment by cowtowncoder** [comment]
I am also hoping to work on figuring out the underlying issue. I created a test:

https://github.com/FasterXML/jackson-core/pull/1265

which suggests some abnormal temporary growth in retention, but not quite the smoking gun. So hoping to learn if there is something specific that might trigger or amplify issues.

Another thing that would be good to know -- although I realize it may not be easy to try it out -- is whether use of one of alternate new pools, `JsonRecyclerPools.newConcurrentDequePool()` -- would show good performance without requiring use of `ThreadLocal`. The reason I am interested in this choice is that for Jackson 2.17.1 there is need to change the default and choices are reverting back to pre-2.17 `ThreadLocal`-pool or `concurrentDequePool`. And it would be great to base decision on observed performance from a real concurrency-heavy use case which I think Trino would qualify as.

**Comment by marvin-roesch** [comment]
@wendigo This situation has improved for us now that we're on 445 with the explicit `ThreadLocal`pool setting. Unfortunately, however, there still appear to places where an `ObjectMapper` is created without any additional settings, which still surfaces this problem, albeit at a much slower pace.

A heap dump I took indicates [`JsonInputFunctions`](https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/operator/scalar/json/JsonInputFunctions.java#L59) is the majority contributor this time around.

Jackson 2.17.1 has been released with a revert to the old pool as default, are there plans to incorporate that into airlift and an upcoming Trino release?
